### ISSUE TITLE
Menu with input suggestions for the catalogs

### ISSUE THREAD

**Issue by sebastian-nehrdich** [ayya-vimala/pali-networks#71]
Currently it is not very convenient to browse to a specific text in our menus as one has to find the right collection/subcollection all the time. Orna suggested to have some kind of a search function where one enters the name/number of a specific text and can immediately jump to the text. 
My first intuition on this would be to use something like this: 
https://www.webcomponents.org/element/Juravenator/paper-dropdown-input
And just load the entire corpus into that. So if one types 4064 this text will be suggested immediately. 
Looks straight forward to implement in my eyes, I might give it a shot the next days.

**Comment by sebastian-nehrdich** [comment]
Orna would prefer this over our current menus. I will see if I can build a prototype and test it a bit. I am afraid that this menutype will not deal too well with long lists of files, but let's see!

**Comment by sebastian-nehrdich** [comment]
I just found this: https://vaadin.com/components/vaadin-combo-box 
Seems easy to implement. Orna and me agree that is better to replace the foldable current menus with a simple searchable one-layer drop-down menu (no folder structure). Folder-structure is served by the visualizations anyway... I will now have a look how we can implement these dropdowns nicely.

**Comment by sebastian-nehrdich** [comment]
I am almost done with this, will upload the alternative menus into a new branch today.

**Comment by sebastian-nehrdich** [comment]
I have now uploaded the version with the dropdown menus to the branch menu-test. 
To note:
1. Currently after selecting the text, the opening of the file does not work. I suppose we need to evoke updateTask at this point, but I didn't figure out yet how to pass the correct parameters here. Maybe it's easy for you Vimala? 
2. Sanskrit menu data is still missing. Will add that the next days.

**Comment by ayya-vimala** [comment]
Maybe it will be good if we have regular meetings so these issues can be discussed as I had not seen it before. I do not get notifications of new issues. Right now I'm struggling with the highlighting so that might take a little while and then I will look into it.
The SC sidemenu is something I want to implement after the database has been installed as it cannot work without that.

**Comment by sebastian-nehrdich** [comment]
Yes no problem! I think we can do our meeting next week Tuesday/Wednesday
(whatever is more convenient) again.
This is all now working as it should so nothing too big to worry about, but
I had to make quite some changes to code and data, so integrating this into
the main branch might take some time.

**Comment by ayya-vimala** [comment]
I would rather see what you want it to look like and do and then build it .. 
this way will be just impossible for me to spend so much time bugfixing because it just ends up with a very hacky solution.

**Comment by sebastian-nehrdich** [comment]
Take my code as a suggestion, we don't have to do it that way! 
It's just that Orna made a bit of pressure to change the menus because she thinks it is quite important at the moment, that's why I went ahead. I also agree that things should not become too hacky, this is why I tried to be careful regarding the quality of the changes I made.

**Comment by ayya-vimala** [comment]
I had a look at the code. It will clash in a few places with the new highlighting code but nothing too difficult to fix. The code looks generally very good. Good idea to use a combo-box.
I also did not like the accordion-layout and agree that this looks better but I'm also worried about having 7000 files in a dropdown! (btw, they need sorting and the text that appears in the dropdown thereafter needs to be cleaned up). I think for the pali-files this would become rather difficult because they are split up into various chapters so you have masses of them. That is one reason why at SC things have been split up in the menus because the dropdowns became much too large to navigate.
I think in the long run a good structured menu would be better, especially for people who are not professionals like Orna and do not know their way around that well. For instance if I want to look at the Tengyur Abhidhamma and I have no idea where that is I will have to search through 4000 files until I find it.

On SC we use elasticsearch as a search engine, but that is rather complicated. I would like that too in the long run because I would be able to search all the files for specific words (like pandaka, that I did my research on last month). That also needs a database.
Another option I had is to have a file-search in that if you know the number, you can go straight there. Something like the filter: you key in "T06TD4021E" and it takes you there.

The only thing I would like is to discuss things a little bit better beforehand so we both know what we are doing and don't clash. I understand that you also have to listen to Orna because she is the boss.

**Comment by ayya-vimala** [comment]
I made a pull request but please don't do anything. I just wanted to merge master into your branch to see if that would clash and what other things need to be done before merging. So as long as I'm working on that branch, please don't do anything about it.

**Comment by ayya-vimala** [comment]
I've added the menu with some changes.
The only thing I do not like about the combo-box is that if I start it up and want to scroll to a sutta it takes a long time and in the mean time I get a white box.  I prefer the dropdowns because at least for me things are quicker to find.
But there will always be some problem with that.
Note that I also changed the values in the tibetan menu json slightly so it looks better.

But I uploaded it to the main site so have a look.
The Chinese still needs work because:
1. I need to have the collection data for the collection filter.
2. some things are broken. Will do that another day because it is 40C down here and I'm melting. :-)
Will make a new issue for point 2.

**Comment by ayya-vimala** [comment]
And I forgot ... would you like to have a simple field that says: type your filenumber here and I will load that text for you ...? That would be very easy to do.

**Comment by sebastian-nehrdich** [comment]
Great! Thanks. I didn't care much about how it looks, so changes to the display are welcome. Also I will probably get new catalogs for the Tibetan soon, then we have to regenerate the menus once again. 
Feel free to push it to the server if you want! 
The combo-box only works well if you know what you want to look for, I agree. Both approaches have their advantages and disadvantages. I saw in the documentation that lazy loading is possible, but I don't know if it will really help. Ideally we would provide both, so having the side menus would also be nice... But as you mentioned elsewhere, we might have to wait until after the DB is there to integrate them. 
Your last comment I don't really understand, do you want to implement a text box as a search function or change something on the combo-box?

And sorry for 'intrusion into your space', I can understand that it was confusing. I thought you might have seen the issue and didn't react because you have been busy with other things, that's why I went ahead. Should send you a mail and inform you about issues that concern directly what you are doing in the future? Sometimes something might come up at a point when we don't have a meeting scheduled or the next meeting is still far away, so it might not always be possible to discuss it beforehand...

**Comment by ayya-vimala** [comment]
I have uploaded it to the server with the addition of a search field so you just type in the ID of the file and there it is! This is mostly for my own use because I find it vastly difficult to find any file right now (I have a limited number of testfiles and they are not on top of the list!). But it is a whole lot quicker to load the site without doing any routing.

I know I'm not always the easiest person to work with! So apologies if I have upset you in any way. I find it a bit overwhelming sometimes when so many changes happen at once and not enough hours in the day to deal with them all. If you want a reaction from me, please do @ayya-vimala in the post or just email to make me aware there is a new issue you want feedback on. 

When I worked with SC/STX we had daily meetings just for such purposes. (I'm not suggesting we should) They were also very strict on coding and Hubert drilled me relentlessly :-) to make make sure I used the right consistent coding everywhere i.e. correct number of spaces, let instead of var, specific notation for strings, ; everywhere, etc. The number of times my pull requests were rejected for such things, even though the code worked. Slowly I began to see the reason for that if you are working with other people and if the code can be pulled by others also.

**Comment by sebastian-nehrdich** [comment]
i totally understand this! I also apologize that I am sometimes just a bit focussed on 'my stuff' and not taking enough time to discuss things thoroughly. It's my first real team project (and first time web development) so I am sorry if I am not always meeting the expectations here, but I try my best! 
I think we have gone a far way and I guess you must have poured already several hundred hours into this and I really appreciate that. Given the fact that we managed to pull more than 40gb of Chinese data into this interface in less than one week without experiencing a shipbreak is testimony to the code quality and robustness of the application. Sure there are problems here and there but we have a clear plan how to tackle them and what next steps we need to take.
Shall we schedule a meeting for next week? On Tuesday, I have a long research meeting in the evening so our usual time (19:00 Japan, 12:00 noon CET), wouldn't work too well. We could try later (like 14-15pm CET) or move it to Wednesday (I have nothing scheduled on Wednesday so that  would be easier). 
To get back on topic: I do feel that the menus are just a bit slow to handle at the moment, probably because there is just way too much data inside of them. I could split Tibetan data into KG and TG and Chinese into X and T which would effectively cut the size of each menu into half. I think that could help a bit. 
Also Orna and me had a lot of discussions regarding the situation of the KG the last days and I will generate new data and menus for all the Tibetan soon. The plan is to strip the menus down to just the Derge number, because that is the data we provide (ACIP is based on Derge). So I will also send new menu data soon.

**Comment by ayya-vimala** [comment]
Maybe we can discuss this. I can easily make TG and KG comboboxes, etc and split that out. Won't take too long. But if you are going to change things to other numbering systems maybe we have to have a good plan first.

I think we have already done a whole lot, far more than we expected. But Orna also needs to understand that if you throw 100.000 and a team of developers at it for 6 months, you can do a whole lot more than we can do together. We can't do everything.

Bhante Sujato also had to learn this the hard way. He believed that me and Blake (who is a professional developer employed fulltime by SC and used to work at Firaxis as a game programmer for Civilization) could develop the SC website within 6 months. So lots of money and 2x 6 months with an extra team of 3 full time professional developers (next ot the two of us) further, he knows better.

**Comment by sebastian-nehrdich** [comment]
I think the changes to the data are not going to be dramatic. Just Tibetan will be reduced to the Derge and possibly Lhasa numbers and there will be no Sanskrit titles, that's all. I think that will also help a bit with the speed.
I agree that we should wait with further ideas for dramatic changes (especially regarding the highlighting) until we all have a clear plan how the solution should look like. 
And I agree that some things only become really possible if you hire the appropiate specialists, I have the impression we have both stretched our expertise to the furthest  and this is about what we can achieve with our resources at the moment...
